objax/variable.py

```diff
@@ -53,15 +53,44 @@ def value(self) -> JaxArray:
     def value(self, tensor: JaxArray):
         raise NotImplementedError('Pure method')
 
-    def assign(self, tensor: JaxArray):
+    def assign(self, tensor: JaxArray, check=True):
         """Sets the value of the variable."""
+        if check:
+            self.assert_assigned_type_and_shape_match(tensor)
         self.value = tensor
 
     def reduce(self, tensors: JaxArray):
         """Method called by Parallel and Vectorize to reduce a multiple-device (or batched in case of vectoriaation)
         value to a single device."""
         if self._reduce:
-            self.assign(self._reduce(tensors))
+            self.assign(self._reduce(tensors), check=False)
+
+    def assert_assigned_type_and_shape_match(self, tensor):
```

Comment:
1. I think we should create a submodule of `objax.util` called `objax.util.check` because this is a lot of checking code.
2. We should name the function `check_assign_type_and_shape(target, source)`

Reply:
Done